OSDOCS#10725 documenting RHEL 9 FIPS requirement (#76748)
openshift/openshift-docs: installing/installing-fips.adoc

For more information about configuring FIPS mode on {op-system-base}, see link:https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/security_hardening/assembly_installing-the-system-in-fips-mode_security-hardening[Installing the system in FIPS mode].
====

For the {op-system-first} machines in your cluster, this change is applied when the machines are deployed based on the status of an option in the `install-config.yaml` file, which governs the cluster options that a user can change during cluster deployment. With {op-system-base-full} machines, you must enable FIPS mode when you install the operating system on the machines that you plan to use as worker machines.

Because FIPS must be enabled before the operating system that your cluster uses boots for the first time, you cannot enable FIPS after you deploy a cluster.

[id="installation-obtaining-fips-installer-oc_{context}"]

> == Obtaining a FIPS-capable installation program using `oc adm extract`

Reviewer: Each of these _Obtaining..._ sections should be in separate modules. They are different user stories. See the [mod docs docs](https://redhat-documentation.github.io/modular-docs/#con-creating-procedure-modules_writing-mod-docs).

> {product-title} requires the use of a FIPS-capable installation binary to install a cluster in FIPS mode. You can obtain this binary by extracting it from the release image using `oc`. After you have obtained the binary, you proceed with the cluster installation, replacing all instances of the `openshift-install` command with `openshift-install-fips`.

Reviewer: ```suggestion
{product-title} requires the use of a FIPS-capable installation binary to install a cluster in FIPS mode. You can obtain this binary by extracting it from the release image by using the {oc-first}. After you have obtained the binary, proceed with the cluster installation, replacing all instances of the `openshift-install` command with `openshift-install-fips`.
```

.Prerequisites

> * You have installed the `oc` command line utility with version 4.16 or newer.

Reviewer: ```suggestion
* You have installed the {oc-first} with version 4.16 or newer.
```

.Procedure

. Extract the FIPS-capable binary from the installation program by running the following command:
+
[source,terminal]
----
$ oc adm release extract --registry-config "${pullsecret_file}" --command=openshift-install-fips --to "${extract_dir}" ${RELEASE_IMAGE}
----
+
where:
+
--
`<pullsecret_file>`:: Specifies the name of a file that contains your pull secret.
`<extract_dir>`:: Specifies the directory where you want to extract the binary.
`<RELEASE_IMAGE>`:: Specifies the Quay.io URL of the {product-title} release you are using. For more information on finding the release image, see xref:../installing/installing_bare_metal_ipi/ipi-install-installation-workflow.adoc#retrieving-the-openshift-installer_ipi-install-installation-workflow[Extracting the OpenShift Container Platform installation program].

> --

Reviewer: No link in a module. Unless you know something that I don't (quite likely!).

Reviewer: Thank you. I noticed that the link was indeed in an assembly. Sorry for doubting you. I think because it was in a procedure, I knee-jerk assumed it was a module!

. Proceed with cluster installation, replacing all instances of the `openshift-install` command with `openshift-install-fips`.

[id="installation-obtaining-fips-installer-mirror_{context}"]
== Obtaining a FIPS-capable installation program using the public OpenShift mirror

> {product-title} requires the use of a FIPS-capable installation binary to install a cluster in FIPS mode. You can obtain this binary by downloading it from the public OpenShift mirror. After you have obtained the binary, you proceed with the cluster installation, replacing all instances of the `openshift-install` binary with `openshift-install-fips`.

Reviewer: ```suggestion
{product-title} requires the use of a FIPS-capable installation binary to install a cluster in FIPS mode. You can obtain this binary by downloading it from the public OpenShift mirror. After you have obtained the binary, proceed with the cluster installation, replacing all instances of the `openshift-install` binary with `openshift-install-fips`.
```

.Prerequisites

* You have access to the internet.

.Procedure

> . Download the installation program from https://mirror.openshift.com/pub/openshift-v4/clients/ocp/latest-4.16/openshift-install-rhel9-amd64.tar.gz.

Reviewer: I am getting a File Not Found error at this link. Is that expected at this point, as 4.16 isn't out yet?

Author: That's correct. right now the file is available in "candidate-4.16" but it will be available in "latest-4.16" upon GA.

====

For the {op-system-first} machines in your cluster, this change is applied when the machines are deployed based on the status of an option in the `install-config.yaml` file, which governs the cluster options that a user can change during cluster deployment. With {op-system-base-full} machines, you must enable FIPS mode when you install the operating system on the machines that you plan to use as worker machines.

Because FIPS must be enabled before the operating system that your cluster uses boots for the first time, you cannot enable FIPS after you deploy a cluster.

[id="installation-obtaining-fips-installer-oc_{context}"]
== Obtaining a FIPS-capable installation program using `oc adm extract`

{product-title} requires the use of a FIPS-capable installation binary to install a cluster in FIPS mode. You can obtain this binary by extracting it from the release image using `oc`. After you have obtained the binary, you proceed with the cluster installation, replacing all instances of the `openshift-install` command with `openshift-install-fips`.

.Prerequisites

* You have installed the `oc` command line utility with version 4.16 or newer.

> * You have downloaded the installation program.

Reviewer 2: Seems like this(`the installation program`) is not required for Obtaining a FIPS-capable installation program, as we'll only use `oc` here

.Procedure

. Extract the dynamically-linked binary from the installation program by running the following command:
+
[source,terminal]
----

> $ oc adm release extract --command=openshift-installer-fips

Author: OK, I've updated the procedure with the correct command (and a link to the BM IPI docs to keep the procedure from repeating info)

----
+
. Proceed with cluster installation, replacing all instances of the `openshift-install` command with `openshift-installer-fips`.

[id="installation-obtaining-fips-installer-mirror_{context}"]
== Obtaining a FIPS-capable installation program using the public OpenShift mirror

{product-title} requires the use of a dynamically-linked installation binary to install a cluster in FIPS mode. You can obtain this binary by downloading it from the public OpenShift mirror. After you have obtained the dynamically-linked binary, you proceed with the cluster installation, replacing all instances of the `openshift-install` binary with `openshift-baremetal-installer`.

.Prerequisites

* You have access to the internet.

.Procedure

> . Download the `openshift-install-rhel9-amd64.tar.gz` file from https://mirror.openshift.com/pub/openshift-v4/clients/ocp/4.16/.

Reviewer 3: ```suggestion
. Download the `openshift-install-rhel9-amd64.tar.gz` file from https://mirror.openshift.com/pub/openshift-v4/clients/ocp/latest-4.16/.
```

Maybe we should just link to the actual file rather than a directory?

. Extract the dynamically-linked installation program. For example, on a computer that uses a Linux operating system, run the following command:
+
[source,terminal]
----
$ tar -xvf openshift-install-rhel9-amd64.tar.gz
----
+
. Copy the `openshift-baremetal-install` binary into the directory that contains your `install-config.yaml` file by running the following command (directory structure is only provided as an example):
+
[source,terminal]
----

> $ cp artcd_working/to_mirror/openshift-v4/x86_64/clients/ocp/4.16.0-rc.3/openshift-baremetal-install ~/install-directory/

Reviewer 3: ick, I didn't realise we did such a poor job of building this tarfile.

Reviewer 3: Proposed fix: https://github.com/openshift-eng/art-tools/pull/700

Reviewer 3: That merged, so we should be able to skip this step now.

{product-title} requires the use of a FIPS-capable installation binary to install a cluster in FIPS mode. You can obtain this binary by downloading it from the public OpenShift mirror. After you have obtained the binary, you proceed with the cluster installation, replacing all instances of the `openshift-install` binary with `openshift-baremetal-install`.

.Prerequisites

* You have access to the internet.

.Procedure

. Download the installation program from https://mirror.openshift.com/pub/openshift-v4/clients/ocp/latest-4.16/openshift-install-rhel9-amd64.tar.gz.
. Extract the installation program. For example, on a computer that uses a Linux operating system, run the following command:
+
[source,terminal]
----
$ tar -xvf openshift-install-rhel9-amd64.tar.gz
----
+
. Copy the `openshift-baremetal-install` binary into the directory that contains your `install-config.yaml` file by running the following command (directory structure is only provided as an example):
+
[source,terminal]
----
$ cp artcd_working/to_mirror/openshift-v4/x86_64/clients/ocp/4.16.0-rc.3/openshift-baremetal-install ~/install-directory/
----
+

> . Proceed with cluster installation, replacing all instances of the `openshift-install` command with `openshift-baremetal-install`.

Reviewer 3: Above fix means we can do:
```suggestion
. Proceed with cluster installation, replacing all instances of the `openshift-install` command with `openshift-install-fips`.
```

|Attributes
|Limitations

|FIPS support in {op-system-base} 9 and {op-system} operating systems.

> .4+|The FIPS implementation does not offer a single function that both computes hash functions and validates the keys that are based on that hash. This limitation will continue to be evaluated and improved in future {product-title} releases.

Author: who would know about this for RHEL 9? since our GA is coming up, would it be a safer bet to leave it in or take it out?

> .4+|The FIPS implementation does not offer a single function that both computes hash functions and validates the keys that are based on that hash. This limitation will continue to be evaluated and improved in future {product-title} releases.

Reviewer 3: Sounds like, whatever this means, it still applies so we can leave it.

> .4+|The FIPS implementation does not offer a single function that both computes hash functions and validates the keys that are based on that hash. This limitation will continue to be evaluated and improved in future {product-title} releases.

Reviewer 2: adding @xiaojiey here to see if she knows something about the current FIPS limitation in OCP clusters.

> .4+|The FIPS implementation does not offer a single function that both computes hash functions and validates the keys that are based on that hash. This limitation will continue to be evaluated and improved in future {product-title} releases.

Reviewer 4: adding the PM @what-do-i-know for clarification

> .4+|The FIPS implementation does not offer a single function that both computes hash functions and validates the keys that are based on that hash. This limitation will continue to be evaluated and improved in future {product-title} releases.

Reviewer 4: adding @neverpanic for more info. I got some feedback in the slack thread https://redhat-external.slack.com/archives/CQ7BBRNQN/p1718781230223409. However, I am not sure how to update it.

> .4+|The FIPS implementation does not offer a single function that both computes hash functions and validates the keys that are based on that hash. This limitation will continue to be evaluated and improved in future {product-title} releases.

Reviewer 5: ```suggestion
.4+|The FIPS implementation does not use a function that performs hash computation and signature generation or validation in a single step. This limitation will continue to be evaluated and improved in future {product-title} releases.
```
I'd propose this alternative, which I think is clearer.
